Fix tests on Windows

From #97, we encountered a few bugs in the tests for Windows. #98 has the mocked and live test steps disabled for the Windows CI. 

1. There are some parts of the tests where local paths are used to write `CloudPath` paths, that need to go through a `PurePosixPath` conversion. See the diff in #97 for those places. 

2. For the live tests, S3 rig setup was failing to upload test assets to the test bucket. It appears to be some kind of authentication problem. Environment variables appear to be generally working, and read/write in a separate script outside the tests had no problems. See some of the changes in #98 to see what was tried. 

```
E           boto3.exceptions.S3UploadFailedError: Failed to upload D:\a\cloudpathlib\cloudpathlib\tests\assets\dir_0\file0_0.txt to ***/TQxxmXB3Vr5dQCLcSgaHxB-test_s3_specific-test_s3path_properties/dir_0/file0_0.txt: An error occurred (SignatureDoesNotMatch) when calling the PutObject operation: The request signature we calculated does not match the signature you provided. Check your key and signing method.
```

(from [this build](https://github.com/drivendataorg/cloudpathlib/runs/1247915278?check_suite_focus=true))

3. There are potentially some issues with the mocked backends. It's possible they overlap with above issues (see 1 above). See [this build's logs](https://github.com/drivendataorg/cloudpathlib/runs/1228827652?check_suite_focus=true)

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -37,7 +37,6 @@ jobs:
       - name: Run mocked tests
         run: |
           make test
-        if: matrix.os != 'windows-latest'
 
       - name: Run live tests
         run: |
@@ -48,7 +47,7 @@ jobs:
           LIVE_S3_BUCKET: ${{ secrets.LIVE_S3_BUCKET }}
           AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
-        if: matrix.python-version == 3.8 && matrix.os != 'windows-latest'
+        if: matrix.python-version == 3.8
 
       - name: Build distribution and test installation
         shell: bash

--- a/cloudpathlib/azure/azblobclient.py
+++ b/cloudpathlib/azure/azblobclient.py
@@ -95,7 +95,12 @@ class AzureBlobClient(Client):
         )
 
         download_stream = blob.download_blob()
-        Path(local_path).write_bytes(download_stream.readall())
+
+        local_path = Path(local_path)
+
+        local_path.parent.mkdir(exist_ok=True, parents=True)
+
+        local_path.write_bytes(download_stream.readall())
 
         return local_path
 

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -2,6 +2,8 @@
 
 black
 flake8
+ipython
+jupyter
 matplotlib
 mkdocs>=1.1
 mkdocs-jupyter
@@ -14,6 +16,7 @@ pytest
 pytest-cases
 pytest-cov
 python-dotenv
+pywin32; sys_platform == 'win32'
 shortuuid
 tabulate
 wheel

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,5 +1,5 @@
 import os
-from pathlib import Path
+from pathlib import Path, PurePosixPath
 
 from azure.storage.blob import BlobServiceClient
 import boto3
@@ -83,7 +83,8 @@ def azure_rig(request, monkeypatch, assets_dir):
         ]
         for test_file in test_files:
             blob_client = blob_service_client.get_blob_client(
-                container=drive, blob=f"{test_dir}/{test_file.relative_to(assets_dir)}"
+                container=drive,
+                blob=str(f"{test_dir}/{PurePosixPath(test_file.relative_to(assets_dir))}"),
             )
             blob_client.upload_blob(test_file.read_bytes(), overwrite=True)
     else:
@@ -126,7 +127,10 @@ def s3_rig(request, monkeypatch, assets_dir):
             f for f in assets_dir.glob("**/*") if f.is_file() and f.name not in UPLOAD_IGNORE_LIST
         ]
         for test_file in test_files:
-            bucket.upload_file(str(test_file), f"{test_dir}/{test_file.relative_to(assets_dir)}")
+            bucket.upload_file(
+                str(test_file),
+                str(f"{test_dir}/{PurePosixPath(test_file.relative_to(assets_dir))}"),
+            )
     else:
         # Mock cloud SDK
         monkeypatch.setattr(

--- a/tests/mock_clients/mock_azureblob.py
+++ b/tests/mock_clients/mock_azureblob.py
@@ -1,5 +1,5 @@
 from datetime import datetime
-from pathlib import Path
+from pathlib import Path, PurePosixPath
 import shutil
 from tempfile import TemporaryDirectory
 
@@ -114,15 +114,15 @@ def mock_item_paged(root, name_starts_with=None):
             and f.is_file()
             and (root / name_starts_with) in [f, *f.parents]
         ):
-            items.append(f)
+            items.append((PurePosixPath(f), f))
 
-    for ix in items:
+    for mocked, local in items:
         # BlobProperties
         # https://github.com/Azure/azure-sdk-for-python/blob/b83018de46d4ecb6554ab33ecc22d4c7e7b77129/sdk/storage/azure-storage-blob/azure/storage/blob/_models.py#L517
         yield BlobProperties(
             **{
-                "name": str(ix.relative_to(root)),
-                "Last-Modified": datetime.fromtimestamp(ix.stat().st_mtime),
+                "name": str(mocked.relative_to(PurePosixPath(root))),
+                "Last-Modified": datetime.fromtimestamp(local.stat().st_mtime),
                 "ETag": "etag",
             }
         )

--- a/tests/mock_clients/mock_s3.py
+++ b/tests/mock_clients/mock_s3.py
@@ -1,6 +1,6 @@
 import collections
 from datetime import datetime
-from pathlib import Path
+from pathlib import Path, PurePosixPath
 import shutil
 from tempfile import TemporaryDirectory
 
@@ -68,7 +68,7 @@ class MockBoto3Object:
             # same file, touch
             self.path.touch()
         else:
-            self.path.write_bytes((self.root / Path(CopySource["Key"])).read_bytes)
+            self.path.write_bytes((self.root / Path(CopySource["Key"])).read_bytes())
 
     def download_file(self, to_path):
         to_path = Path(to_path)
@@ -123,7 +123,11 @@ class MockObjects:
 
     def filter(self, Prefix=""):
         path = self.root / Prefix
-        items = [f for f in path.glob("**/*") if f.is_file() and not f.name.startswith(".")]
+        items = [
+            PurePosixPath(f)
+            for f in path.glob("**/*")
+            if f.is_file() and not f.name.startswith(".")
+        ]
         return MockCollection(items, self.root)
 
 
@@ -145,8 +149,8 @@ class MockCollection:
 
     def delete(self):
         for p in self.full_paths:
-            p.unlink()
-            delete_empty_parents_up_to_root(p, self.root)
+            Path(p).unlink()
+            delete_empty_parents_up_to_root(Path(p), self.root)
 
         return [{"ResponseMetadata": {"HTTPStatusCode": 200}}]
 

--- a/tests/test_cloudpath_file_io.py
+++ b/tests/test_cloudpath_file_io.py
@@ -1,5 +1,6 @@
 from datetime import datetime
 import os
+from pathlib import PurePosixPath
 from time import sleep
 
 import pytest
@@ -80,9 +81,12 @@ def test_file_read_writes(rig, tmp_path):
     dl_dir.mkdir(parents=True, exist_ok=True)
     p3.download_to(dl_dir)
     cloud_rel_paths = sorted(
-        [p._no_prefix_no_drive[len(rig.test_dir) + 1 :] for p in p3.glob("**/*")]
+        # CloudPath("prefix://drive/dir/file.txt")._no_prefix_no_drive = "/dir/file.txt"
+        [p._no_prefix_no_drive[len(rig.test_dir) + 2 :] for p in p3.glob("**/*")]
     )
-    dled_rel_paths = sorted([str(p)[len(str(dl_dir)) :] for p in dl_dir.glob("**/*")])
+    dled_rel_paths = sorted(
+        [str(PurePosixPath(p.relative_to(dl_dir))) for p in dl_dir.glob("**/*")]
+    )
     assert cloud_rel_paths == dled_rel_paths
 
 

